alias.c

```diff
@@ -83,20 +83,29 @@ static void expand_aliases_r(struct AddressList *al, struct ListHead *expn)
           }
         }
 
-        if (!i)
+        if (i)
         {
-          mutt_list_insert_head(expn, mutt_str_strdup(a->mailbox));
-          struct AddressList copy = TAILQ_HEAD_INITIALIZER(copy);
-          mutt_addrlist_copy(&copy, alias, false);
-          expand_aliases_r(&copy, expn);
-          struct Address *a2 = NULL, *tmp = NULL;
-          TAILQ_FOREACH_SAFE(a2, &copy, entries, tmp)
-          {
-            TAILQ_INSERT_BEFORE(a, a2, entries);
-          }
-          a = TAILQ_PREV(a, AddressList, entries);
-          TAILQ_REMOVE(al, TAILQ_NEXT(a, entries), entries);
+          // We've already seen this alias, so drop it
+          struct Address *next = TAILQ_NEXT(a, entries);
+          TAILQ_REMOVE(al, a, entries);
+          mutt_addr_free(&a);
+          a = next;
+          continue;
+        }
+
+        mutt_list_insert_head(expn, mutt_str_strdup(a->mailbox));
+        struct AddressList copy = TAILQ_HEAD_INITIALIZER(copy);
+        mutt_addrlist_copy(&copy, alias, false);
+        expand_aliases_r(&copy, expn);
```

Comment:
Can you add a comment about what's happening here?

From what I can put together, this is the recursive expansion of aliases to ensure nested aliases work. Is that correct?